### PR TITLE
Caches Maven repository that Docker uses

### PR DESCRIPTION
**Pull request by adriancole**
Maven build commands operate inside Docker. Before, we had no means to
cache the localRepository used for each project due to this. This
meant each build would re-download all dependencies. This didn't
dominate execution time for reasons including our Docker parent image
must be remotely fetched. In other words, downloads from Maven central
happened fast enough to not be noticed.

This uses a directory `m2repository` as a rendezvous between what CI
can cache and what Docker can read: CI feeds this (via `configure_*`)
and docker reads this (via .dockerignore). When the build executes,
anything found in `m2repository` copies into `~/.m2/repository`,
obviating remote downloads.

This also removes the useless and distracting Docker cache statements.

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -36,15 +36,11 @@ jobs:
...
-          key: ${{ runner.os }}-${{ matrix.version }}-maven-${{ hashFiles('**/pom.xml') }}
-          restore-keys: ${{ runner.os }}-${{ matrix.project }}-maven-
-      - name: Cache Docker
-        uses: actions/cache@v2
-        with:
-          path: ~/.docker
-          key: ${{ runner.os }}-${{ matrix.version }}-docker-${{ hashFiles('**/Dockerfile') }}
-          restore-keys: ${{ runner.os }}-${{ matrix.project }}-docker
+          path: ./m2repository  # Shared with the Docker build context via .dockerignore
+          key: ${{ runner.os }}-${{ matrix.version }}-maven-${{ hashFiles('parent-pom.xml', format('{0}/pom.xml', matrix.version)) }}
```
Comment (adriancole) [inline .github/workflows/deploy.yml:40]: fyi @openzipkin/devops-tooling if you ever wanted to know.. nested variables aren't supported in GH actions. You have to play games that feel like awk :D

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -36,15 +36,11 @@ jobs:
       - name: Cache local Maven repository
         uses: actions/cache@v2
         with:
-          path: ~/.m2/repository
-          key: ${{ runner.os }}-${{ matrix.version }}-maven-${{ hashFiles('**/pom.xml') }}
-          restore-keys: ${{ runner.os }}-${{ matrix.project }}-maven-
-      - name: Cache Docker
```
Comment (adriancole) [inline .github/workflows/deploy.yml:42]: this was completely useless. see https://dev.to/dtinth/caching-docker-builds-in-github-actions-which-approach-is-the-fastest-a-research-18ei for explanation

File: .gitignore [hunk 3]
```diff
@@ -6,3 +6,8 @@ target/
 *.iml
 .idea
 zipkin.jar
+
+# We cannot share paths outside the current directory with the Docker context
+# (ex via symlinks). So, build-bin/maven_go_offline overrides localRepository.
+m2repository/
+.settings.xml
```
Comment (adriancole) [inline .gitignore:13]:
```suggestion
```

File: .dockerignore [hunk 4]
```diff
@@ -21,3 +21,6 @@
 !**/pom.xml
 !/parent-pom.xml
 
+# We cannot share paths outside the current directory with the Docker context
+# (ex via symlinks). So, build-bin/maven_go_offline overrides localRepository.
```
Comment (adriancole) [inline .dockerignore:25]:
```suggestion
# (ex via symlinks). So, build-bin/configure_* overrides maven.repo.local
```

File: .gitignore [hunk 5]
```diff
@@ -6,3 +6,7 @@ target/
 *.iml
 .idea
 zipkin.jar
+
+# We cannot share paths outside the current directory with the Docker context
+# (ex via symlinks). So, build-bin/maven_go_offline overrides localRepository.
```
Comment (adriancole) [inline .gitignore:11]:
```suggestion
# (ex via symlinks). So, build-bin/configure_* overrides maven.repo.local
```

**Review by adriancole — COMMENTED** [review]
drift

File: docker/Dockerfile [hunk 6]
```diff
@@ -16,6 +16,9 @@ FROM ghcr.io/openzipkin/java:11.0.9_p11 as install
 
 WORKDIR /code
 COPY --from=scratch /code .
+# If we've seeded m2repository, copy it into the repository of our install image.
+# This prevents uncacheable downloads that occur during Docker build.
+RUN test -d /code/m2repository && cp -r /code/m2repository/* ~/.m2/repository/
```
Comment (anuraaga) [inline docker/Dockerfile:21]: Is it not possible to point maven at `/code/m2repository` without copying?

**Review by anuraaga — APPROVED** [review]

File: docker/Dockerfile (hunk 6 above)
Comment (adriancole) [reply docker/Dockerfile:21]: yeah I think so. I already monkeyed with the scripts to have them use mvnw so it should work.

Comment (adriancole) [reply docker/Dockerfile]: 🤞

Comment (adriancole) [reply docker/Dockerfile]: it wasn't so easy, but it is better in general now I think.

File: build-bin/maven/maven_go_offline [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+# This is a go-offline that properly works with multi-module builds
+
+set -ue
+
+export MAVEN_OPTS="$($(dirname "$0")/maven_opts)"
+if [ -x ./mvnw ]; then alias mvn=${PWD}/mvnw; fi
+
+(
+  if [ "${MAVEN_PROJECT_BASEDIR:-.}" != "." ]; then cd ${MAVEN_PROJECT_BASEDIR}; fi
+
```
Comment (adriancole) [inline build-bin/maven/maven_go_offline:25]:
```suggestion
```

**Pull request merged**